torch_geometric/utils/_negative_sampling.py

```diff
@@ -289,99 +356,152 @@ def structured_negative_sampling_feasible(
     num_nodes = maybe_num_nodes(edge_index, num_nodes)
     max_num_neighbors = num_nodes
 
-    edge_index = coalesce(edge_index, num_nodes=num_nodes)
-
     if not contains_neg_self_loops:
         edge_index, _ = remove_self_loops(edge_index)
         max_num_neighbors -= 1  # Reduce number of valid neighbors
 
     deg = degree(edge_index[0], num_nodes)
-    # True if there exists no node that is connected to all other nodes.
-    return bool(torch.all(deg < max_num_neighbors))
+    # structured sample is feasible if, for each node, deg > max_neigh/2
+    return bool(torch.all(2 * deg <= max_num_neighbors))
 
 
 ###############################################################################
 
 
-def sample(
-    population: int,
-    k: int,
-    device: Optional[Union[torch.device, str]] = None,
-) -> Tensor:
-    if population <= k:
-        return torch.arange(population, device=device)
-    else:
-        return torch.tensor(random.sample(range(population), k), device=device)
+def get_method(method: str, size: Tuple[int, int]) -> str:
+    # select the method
+    tot_num_edges = size[0] * size[1]
+    # prefer dense method if the graph is small
+    auto_method = 'dense' if tot_num_edges < _MAX_NUM_EDGES else 'sparse'
+    method = auto_method if method == 'auto' else method
 
+    if method == 'dense' and tot_num_edges >= _MAX_NUM_EDGES:
+        warnings.warn(
+            f'You choose the dense method on a graph with {tot_num_edges} '
+            f'possible edges! It could require a lot of memory!')
 
-def edge_index_to_vector(
-    edge_index: Tensor,
-    size: Tuple[int, int],
-    bipartite: bool,
-    force_undirected: bool = False,
-) -> Tuple[Tensor, int]:
+    return method
 
-    row, col = edge_index
 
-    if bipartite:  # No need to account for self-loops.
-        idx = (row * size[1]).add_(col)
-        population = size[0] * size[1]
-        return idx, population
+def sample_almost_k_edges(
+    size: Tuple[int, int],
+    k: Optional[int],
+    force_undirected: bool,
+    remove_self_loops: bool,
+    method: str,
+    device: Optional[Union[torch.device, str]] = None,
+) -> Tuple[Tensor, Tensor]:
 
-    elif force_undirected:
-        assert size[0] == size[1]
-        num_nodes = size[0]
+    assert method in ['sparse', 'dense']
+    N1, N2 = size
 
-        # We only operate on the upper triangular matrix:
-        mask = row < col
-        row, col = row[mask], col[mask]
-        offset = torch.arange(1, num_nodes, device=row.device).cumsum(0)[row]
-        idx = row.mul_(num_nodes).add_(col).sub_(offset)
-        population = (num_nodes * (num_nodes + 1)) // 2 - num_nodes
-        return idx, population
+    if method == 'sparse':
+        assert k is not None
+        k = 2 * k if force_undirected else k
+        p = torch.tensor([1.], device=device).expand(N1 * N2)
+        if k > N1 * N2:
+            k = N1 * N2
+        new_edge_id = torch.multinomial(p, k, replacement=False)
```

Comment:
```suggestion
        assert k is not None
        k = 2 * k if force_undirected else k
        new_edge_id = torch.randperm(N1*N2, device = device)[:min(k, N1*N2)]
```
Why not just do this? Also given this are we saying `dense` and `sparse` are not different anymore. Previosly the difference in dense and sparse was in how negatives were identified, now we don't have that difference.

Reply:
The methods differ in terms of complexity:
- the  __dense__ method enumerates all the possible edges and then filters out the positive ones. This requires $O(N^2)$ space and time.
- the __sparse__ method samples $k$ edges and then filters out the positive ones. The complexity now scales with $O(k)$.

The idea is that if we have a big graph, the dense method is prohibitive. However, in most cases, big graphs are not dense (i.e. $E \ll N^2$). Thus, we can still obtain negative edges by sampling $k$ edges at random since the probability to get a positive edge is low (it is $E/N^2 \ll 1$). Thus, thanks to the sparse method, we can obtain the negative edges without enumerating all the possible edges (as in the dense method).

In your suggestion, both sparse and dense will use `torch.randperm(N1*N2)`, which returns a vector of size N1*N2, i.e. $O(N^2)$.

Reply:
`torch.expand` does not allocate new memory, as you can see in the [pytorch documtation](https://pytorch.org/docs/stable/generated/torch.Tensor.expand.html). 
Actually, I am not sure about the complexity of `torch.multinomial` when `replacement=False`, but I didn't find anything online. However, this can be replaced by `torch.randint` and then handling the repetition of negative edges.